### DISCUSSION TITLE
Questions regarding this project.

### DISCUSSION THREAD

**Discussion by Raj2032** [EmbarkStudios/rust-gpu#842] [General] [answered]
Hey guys I have some questions regarding this game engine:

1. Does it have a "locked in framework" like Bevy does with their ecs model or does it allow you to code freely without being locked into some framework?
2. Is this suitable to make games?
3. Is there a gui game editor available or not yet?
4. Is this for 2D or 3D games or both?
5. Does it do multithreading like Bevy?
6. How does performance compare to something such as Bevy?
7. Does it use wgpu as its backend?
8. Is there some egui implementation for this?
9. Am I able to fork this game engine, modify it, rename it, close source my fork and even sell it provided I provide credits (not that I am planning to do this at all as open source is the way to go but just curious)?

**Comment by khyperia** [comment, marked as answer]
> Hey guys I have some questions regarding this game engine:

Hey, thanks for the questions! I think there's a bit of a misunderstanding here: rust-gpu is not a game engine, but rather a way to program shaders in rust. In that sense, rust-gpu is designed to be used as a shader language for other frameworks or game engines - at the moment, there's not even any opinions around how the CPU side binding model behaves. The usual behavior is that the output of rust-gpu, shader modules, are loaded via vulkan into whatever framework/engine - whether that's `ash`, `wgpu`, or `bevy`.

> * Does it have a "locked in framework" like Bevy does with their ecs model or does it allow you to code freely without being locked into some framework?

Because it's not a game engine, I think this question doesn't really apply - there is no CPU-side code when using rust-gpu, so nothing like ECS or similar.

> * Is this suitable to make games?

Yes! We're using it internally in some of our game projects, which you can read about [here](https://medium.com/embarkstudios/homegrown-rendering-with-rust-1e39068e56a7). I believe there are also efforts to integrate it into bevy, and perhaps some community folks are using it in wgpu/etc.

> * Is there a gui game editor available or not yet?

No GUI or whatever - it's "just" a shading language.

> * Is this for 2D or 3D games or both?

Both 2D and 3D shaders are supported! Whatever a GLSL shader can do, rust-gpu can do too.

> * Does it do multithreading like Bevy?

I guess GPUs are multithreaded? 😛

> * How does performance compare to something such as Bevy?

Apples and oranges! Rust-gpu would be used by something like bevy, not replace it.

> * Does it use wgpu as its backend?

If you'd like, sure! rust-gpu shaders are known to work with wgpu (although naga has quite a few bugs in it - we recommend using wgpu's spirv bypass features). rust-gpu shaders can also be used in ash, and however esoteric wrapper of vulkan or spirv-cross you can think of.

> * Is there some egui implementation for this?

I think it'd be the other way around - egui would use rust-gpu to implement its own rendering functionality! (I don't believe egui uses rust-gpu, though)

> * Am I able to fork this game engine, modify it, rename it, close source my fork and even sell it provided I provide credits (not that I am planning to do this at all as open source is the way to go but just curious)?

rust-gpu is dual licensed under apache-2.0 and MIT (the license pretty standard in the rust community). Please refer to those licenses for usage information!

**Reply by Raj2032** [reply]
Thanks mate :)
